lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+ * transforms those tags by adding, removing, or updating the
+ * values of the HTML attributes within that tag (opener).
+ *
+ * Does not fully parse HTML or _recurse_ into the HTML structure
+ * Instead this scans linearly through a document and only parses
+ * the HTML tag openers.
+ *
+ * @TODO:
+ *  - Write a comprehensive test suite for various unexpected
+ *    quirks and attribute names. Unfortunately we need to have
+ *    some fairly robust support because it's not difficult to
+ *    create misleading HTML through the block editor. For
+ *    example, as can happen with a number of plaintext inputs.
+ *
+ * Performance ideas:
+ *
+ * It may be possible that we're thrashing the garbage collector by
+ * allocating new classes for each tag and attribute token as we
+ * scan through the input document. If this is the case we could
+ * try pre-allocating a certain number of token class instances and
+ * reuse those upon scanning each new tag. This would incur a fixed
+ * memory cost but in addition to lowering the pressure on the
+ * garbage collector we might end up having more efficient use of
+ * localized data within the processor cache.
+ *
+ * If we can pick up scanning where we leave off, we can stop parsing
+ * within a tag opener as soon as we meet all the query constraints.
+ * If we need to set an attribute we haven't yet seen then we would
+ * need to continue scanning to the end of that tag opener until we
+ * find the existing attribute or know it's not there. This could
+ * avoid parsing large attributes, such as one with a long data-URI.
+ *
+ * Defer creating comparable values until we need them so we don't
+ * eagerly allocate and convert string values we don't ever read.
+ *
+ * Classes are passed by reference by default, but are we missing
+ * opportunities to pass by reference when we are copying by value?
+ */
+
+
+/**
+ * Processes an input HTML document by applying a specified set
+ * of patches to that input. Tokenizes HTML but does not fully
+ * parse the input document.
+ */
+class WP_HTML_Processor {
+	/**
+	 * Describes the constraints imposed on the HTML tag we're trying to match.
+	 *
+	 * @see WP_Tag_Find_Descriptor::parse()
+	 *
+	 * @var WP_Tag_Find_Descriptor
+	 */
+	public $descriptor;
+
+
+	/**
+	 * Scans the HTML document, keeping track of its own state.
+	 *
+	 * @var WP_HTML_Scanner
+	 */
+	public $scanner;
+
+
+	/**
+	 * Set of modifications to perform on input HTML document, built
+	 * via calls to the fluid API exposed by WP_HTML_Processor.
+	 *
+	 * @var WP_Tag_Modifications
+	 */
+	public $modifications;
+
+
+	public function __construct( $input ) {
+		$this->scanner = new WP_HTML_Scanner( $input );
+		$this->modifications = new WP_Tag_Modifications();
+	}
+
+
+	public function find( $query ) {
+		$this->commit_class_changes();
+		$this->descriptor = WP_Tag_Find_Descriptor::parse( $query );
+		if ( false === $this->scanner->scan( $this->descriptor ) ) {
+			return false;
```

Comment:
`return false` here breaks the fluid interface idea and forces the consumer to always check the return value. I'm a bit bummed to see calls like `$p->find()->set_attribute()->find()...` go, but I guess it's the only reasonable approach. As a consumer, I wouldn't want to chain a few `find()` calls only to have them all fail because the first one didn't find anything.

Reply:
🤔

Reply:
I'm not completely attached to the fluid interface. I could give it up if we need because I think it's valuable to use a `find()` inside a loop as the termination indicator.

```php
$number = 1;
while ( $p->find( [ 'tag_name' => 'section' ] ) ) {
    $p->set_attribute( 'data-section-id', $number++ );
}
```

Reply:
I believe this is our choice:

* A succinct fluid interface that will sometimes cause unexpected errors
* A slightly more verbose imperative interface that won't have cause unexpected errors

And in this case, I feel comfortable letting go of the fluid interface.

Reply:
> I feel comfortable letting go of the fluid interface.

That seems fine, and while fluid interfaces have their own love, they also have their own controversy. At least for what it's worth I feel like a lot of PHP code follows this idiomatic pattern which looks like a fluid interface without being one.

```php
$x = 'some value';
$x = some_tranform( $x );
$x = another_thing( $x, $y );
$x = with_more_intensity( $x );

$p = new GoGetter();
$p->go_get_ore();
$p->refine_it();
$p->shake_it();
$p->schmelt_it();
$p->dump_it();
```